src/v/config/property.h

```diff
@@ -626,8 +630,12 @@ class retention_duration_property final
     // serialization point in config_store::to_json to avoid users from being
     // forced to consume the property as a json object.
     void to_json(
-      json::Writer<json::StringBuffer>& w, redact_secrets) const final {
-        json::rjson_serialize(w, _value.value_or(-1ms));
+      json::Writer<json::StringBuffer>& w, redact_secrets redact) const final {
+        if (is_secret() && redact == redact_secrets::yes) {
+            json::rjson_serialize(w, secret_placeholder);
```

Comment:
We shouldn't change the type of the output here (it's a time integer).

In practice we will never have secret retention periods, so perhaps just use a `vassert(!is_secret()...` to make it clear that this function doesn't handle redaction but shouldn't have to.